### ISSUE TITLE
Built-in supporting for attaching and displaying external image

### ISSUE THREAD

**Issue by hsiang-wu** [mastodon/mastodon#14349]
### Pitch

So the idea is to have a "first-class" external image support. When a user wants to attach an online image to her toots, she could just paste the url of that image. And when others see the toot, the external image would be displayed as an image not a url. The user could also mark the image NSFW just as one can do so for an uploaded image.

Host should be able to enable / disable the feature.

### Motivation

This could make hosting an instance less expensive.

So there was a sudden increase of new comers to the instance I use, which severely slowed down the server and forced the admins to upgrade the server. After the upgrade, the admins were saying something like "we may not be able to afford upgrading the server in the future at this pace, and we do want to keep open to everybody, so if everyone could try use external links for images (local images could be uploaded to an image hosting website) instead of uploading them to our server, it could really help a lot". But the thing is, currently if I post an url of image in the toot, others could only see the url, and it requires another click to see the image, it's bad user experience thus not many will follow the advice, if that's not a forced policy.

Arguably part of the feature could be implemented as 3rd party app, but marking NSFW would be hard for apps that doesn't support such feature. Thus it might be better to have support external image links as "first-class" way of attaching images. I'd think 3rd party apps support uploading a local image to an image host for the user case, and mastodon supports the core feature of attaching & displaying external images.

**Comment by ThibG** [comment]
One issue with that is privacy: as an end-user, when using Mastodon, you don't ever load anything that is not explicitly approved by the instance admin (except for embedded previews, but they are behind a click-through). Allowing this feature would change that, for instance enabling someone to know some particular user is online by sending them a DM with an external image they control.

**Comment by hsiang-wu** [comment]
I see. Two workarounds I can think of:

1. Require another click-through for all external images (either no preview or preview cached at host when the image is posted). Still better than opening external image links in new windows in terms of user experience;

And / or,

2. Ask admin for a whitelist for this feature. Only images from certain hosts explicitly approved by the admins would be allowed to be posted (strict) / automatically loaded (looser policy, can be combined with the other workaround).

**Comment by trwnh** [comment]
It should be possible to allow users to attach media without uploading it? At least from an API/backend/federation level, I don’t see an issue with it. You would have an `attachment` item just the same, but the `url` of it would be hosted externally.

Re: privacy concerns, doing it this way would not be a problem because the media could be fetched by the instance just like any other media. It falls on servers to evict old media from cache as appropriate.

**Comment by mjankowski** [comment]
Sorry, to clarify my reasoning - I agree with you that it would be completely _possible_ to do; but the previous remark made me think it is not _going to_ be done...
